presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -460,49 +497,38 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
             return getMagicLiteralFunctionSignature(type);
         }
 
-        // TODO this should be made to work for any parametric type
-        FunctionInfo fieldReference = getRowFieldReferenceFunctionInfo(name, parameterTypes);
-        if (fieldReference != null) {
-            return fieldReference.getSignature();
+        // TODO this should be removed and implemented as a special expression type
+        if (!parameterTypes.isEmpty() && parameterTypes.get(0).getBase().equals(StandardTypes.ROW)) {
+            ParametricFunction fieldReference = getRowFieldReference(name.getSuffix(), parameterTypes.get(0));
+            if (fieldReference != null) {
+                return bindSignature(fieldReference.getSignature(), resolvedTypes, true, typeManager);
+            }
         }
 
         throw new PrestoException(FUNCTION_NOT_FOUND, message);
     }
 
-    private FunctionInfo getRowFieldReferenceFunctionInfo(QualifiedName name, List<TypeSignature> parameterTypes)
+    @Nullable
+    private ParametricFunction getRowFieldReference(String field, TypeSignature rowTypeSignature)
     {
-        List<Type> resolvedTypes = resolveTypes(parameterTypes, typeManager);
-        FunctionInfo match = null;
-        for (TypeSignature typeSignature : parameterTypes) {
-            if (typeSignature.getBase().equals(StandardTypes.ROW)) {
-                RowType rowType = RowParametricType.ROW.createType(resolveTypes(typeSignature.getParameters(), typeManager), typeSignature.getLiteralParameters());
-                // search for exact match
-                for (ParametricFunction function : RowParametricType.ROW.createFunctions(rowType)) {
-                    if (!function.getSignature().getName().equals(name.toString())) {
-                        continue;
-                    }
-                    Map<String, Type> boundTypeParameters = function.getSignature().bindTypeParameters(resolvedTypes, false, typeManager);
-                    if (boundTypeParameters != null) {
-                        checkArgument(match == null, "Ambiguous call to %s with parameters %s", name, parameterTypes);
-                        try {
-                            match = specializedFunctionCache.getUnchecked(new SpecializedFunctionKey(function, boundTypeParameters, resolvedTypes.size()));
-                        }
-                        catch (UncheckedExecutionException e) {
-                            throw Throwables.propagate(e.getCause());
-                        }
-                    }
-                }
-
-                if (match != null) {
-                    return match;
-                }
+        Type rowType = typeManager.getType(rowTypeSignature);
+        checkState(rowType.getTypeSignature().getBase().equals(StandardTypes.ROW), "rowType is not a ROW type");
+        ParametricFunction match = null;
+        for (ParametricFunction function : RowParametricType.ROW.createFunctions(rowType)) {
```

Comment:
I find it very weird that we create accessors for all fields and then pick the right one, instead of passing in the name and create just the right accessor.